Extend our type signatures of inner and dot to match NumPy's types.

adamreichold opened the pull request:
Closes #284

adamreichold commented on src/sum_products.rs:
```diff
@@ -1,28 +1,68 @@
+use crate::npyffi::{array::PY_ARRAY_API, NPY_CASTING, NPY_ORDER};
+
+/// Return value of a function that can yield either an array or a scalar.
+pub trait ArrayOrScalar<'py, T>: FromPyObject<'py> {}
```
I chose to go for a trait in return position so that the current scheme of selecting the result type via type inference continues to work. (Compared to say using an `enum`.)

We might want to consider sealing the trait, but since the trait itself is mainly documentation (I could have use a `FromPyObject` bound directly on the methods) and nothing bad will happen if anyone implements this for other types, I decided against the additionally code noisy necessary for sealing.

davidhewitt commented on tests/sum_products.rs:
```diff
@@ -1,11 +1,11 @@
         let a = pyarray![py, [1, 0], [0, 1]];
         let b = pyarray![py, [4, 1], [2, 2]];
-        let c = dot(a, b).unwrap();
+        let c: &PyArray2<_> = dot(a, b).unwrap();
```
To avoid needing to specify the result type for inference, do you think that there's a case for making a `Dot` trait?

Then we have something like `fn dot(x, T1, y: T2) -> <T1 as Dot::<T2>>::Result`, and can define all the return types based on input types.

adamreichold replied:
Is that possible for e.g. dynamically sized arrays? I think NumPy would still return `float64` if the `IxDyn` inputs where e.g. one-dimensional so that at least some cases would not be visible on the type level.

adamreichold replied:
(There is admittedly also my general preference for the ndarray API serving the same purpose, so that I somewhat inclined towards simplicity of implementation here compared to say usability.)

adamreichold replied:
((Finally, I am not even sure that the originally reported problem - getting `PyArray0` for `T=i32` but `f64` for `T=f64` - could be solved on the type-level without specialization?))

adamreichold replied:
> ((Finally, I am not even sure that the originally reported problem - getting PyArray0 for T=i32 but f64 for T=f64 - could be solved on the type-level without specialization?))

It appears that would not be a problem: One can `extract` a `T` from a zero-dimensional `numpy.ndarray` with element type `T`, i.e. returning a scalar should always work if the result is conceptually scalar. The bug report is for the other way around, i.e. one cannot `extract` a `PyArray0<T>` when a bare `T` is actually returned by the NumPy routine.

davidhewitt replied:
Right, yep I understand now. Let's keep as you've designed 👍

kngwyu reviewed (changes requested):
Oops, I didn't notice that this can cause an error, thanks.
Could you please some test cases that extract floats/integers also in (non-doc) tests?

Also, can we do the same for einsum?

adamreichold commented:
> Oops, I didn't notice that this can cause an error, thanks.

Yeah, I think the effective signature of the capsule API is quite unfortunate considering a zero-dimensional array can be used almost exactly like a scalar in Python code.

> Could you please some test cases that extract floats/integers also in (non-doc) tests?

Added more tests for `inner` and `dot` as a separate commit.

> Also, can we do the same for einsum?

Indeed this implies here too depending on the summation performed. Added as a separate commit.

adamreichold commented:
As this break the API anyway, I took the liberty of renaming the `einsum_impl` function to `einsum`. There is no danger of conflicts as macros and functions inhabit distinct namespaces and most people will invoke the macro anyway. But I think the `_impl` suffix makes the name look unnecessary "internal."

adamreichold commented:
Slipped in a further change to make use of `PyArray::item` to basically game our coverage up... |-\

adamreichold commented:
@kngwyu Are you satisfied with the adjustments?

kngwyu reviewed (approved):
Sorry for being late. Thank you for addressing my requests!
